diff --git a/lua/bufferline/buffer.lua b/lua/bufferline/buffer.lua
index 6fd809f..8abec16 100644
--- a/lua/bufferline/buffer.lua
+++ b/lua/bufferline/buffer.lua
@@ -43,7 +43,7 @@ local function get_name(opts, number)
   --- @type nil|string
   local name = buf_is_valid(number) and buf_get_name(number) or nil
 
-  if name then
+  if name and name ~= '' then
     name = buf_get_option(number, 'buftype') == 'terminal' and terminalname(name) or utils.basename(name)
   elseif opts.no_name_title ~= nil and opts.no_name_title ~= vim.NIL then
     name = opts.no_name_title
diff --git a/lua/bufferline/layout.lua b/lua/bufferline/layout.lua
index 22b542b..2ef135c 100644
--- a/lua/bufferline/layout.lua
+++ b/lua/bufferline/layout.lua
@@ -62,8 +62,8 @@ local function calculate_buffers_width(state, base_width)
              or opts.icon_close_tab_modified)
 
         width = width
-          + strwidth(icon)
-          + 1 -- space-after-close-icon
+          + strwidth(icon or '')
+          -- + 1 -- space-after-close-icon
       end
     end
     sum = sum + width
diff --git a/lua/bufferline/render.lua b/lua/bufferline/render.lua
index 42a4fb1..eeb7a6e 100644
--- a/lua/bufferline/render.lua
+++ b/lua/bufferline/render.lua
@@ -323,7 +323,7 @@ local function render(update_names)
         {jumpLetterPrefix,   jumpLetter},
         {clickable .. namePrefix,         name},
         {'',                 padding},
-        {'',                 ' '},
+--        {'',                 ' '},
         {closePrefix,        close},
       }
     }
